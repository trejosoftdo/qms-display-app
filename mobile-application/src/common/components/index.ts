export { default as ActionButton } from './ActionButton';
export { default as ToConfigurationNavigationButton } from './ToConfigurationNavigationButton';
export { default as AppView } from './AppView';
export { default as Value } from './Value';
export { default as ConditionalContainer } from './ConditionalContainer';
export { default as Cards } from './Cards';
export { default as AppHeader } from './AppHeader';
