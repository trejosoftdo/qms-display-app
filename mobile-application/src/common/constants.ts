export const AUTHORIZATION_PENDING_CODE = 'authorization_pending';

export const VALIDATION_ERROR_STATUS_CODE = 400;

export const DEVICE_NOT_CONNECTED_ERROR_MESSAGE = 'Device is not connected';

export const CONNECTION_EXPIRED_ERROR_MESSAGE = 'Connection expired!';

export const APP_PROTO = 'app://';

export const DEVICE_ID_KEY = 'SECURE_DEVICE_ID';

export const CONNECTION_DETAILS_KEY = 'CONNECTION_DETAILS';

export const WEB_PLATFORM = 'web';

export const BEARER_PORTION = 'Bearer';

export const EMPTY_VALUE = '';

export const DEFAULT_REGULAR_FONT_NAME = 'Montserrat-Regular';

export const DEFAULT_LIGHT_FONT_NAME = 'Montserrat-Light';

export const DEFAULT_MEDIUM_FONT_NAME = 'Montserrat-Medium';

export const NORMAL_FONT_WEIGHT = 'normal';

export const ONE_SECOND_MILISECONDS = 1000;

export const INTERVAL_TIME = 10 * ONE_SECOND_MILISECONDS;

export const SERVICES_PATH = '/services';

export const TICKET_DETAILS_PATH = '/ticket-details';

export const HOME_PATH = '/';

export const CATEGORIES_PATH = '/categories';

export const CONFIGURATION_PATH = '/configuration';

export const NO_OP = () => {};


export const LANGUAGES = {
  SPANISH: 'es',
  ENGLISH: 'en',
};

export const CONFIG_OPTIONS = {
  INFORMATION: 'information',
  CONNECTION: 'connection',
  LANGUAGE: 'language',
};
