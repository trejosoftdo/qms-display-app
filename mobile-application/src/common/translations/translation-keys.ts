export const TRANSLATION_CATEGORIES_KEY = 'translation:categories';

export const TRANSLATION_CONFIGURATION_KEY = 'translation:configuration';

export const TRANSLATION_CONNECTION_KEY = 'translation:connection';

export const TRANSLATION_INFORMATION_KEY = 'translation:information';

export const TRANSLATION_SERVICES_KEY = 'translation:services';

export const TRANSLATION_TICKET_DETAILS_KEY = 'translation:turnDetails';

export const TRANSLATION_LANGUAGE_KEY = 'translation:language';

export const TRANSLATION_CUSTOMER_DATA_KEY = 'translation:customerData';

export const TRANSLATION_WAIT_MESSAGE_KEY = 'translation:waitMessage';

export const TRANSLATION_CHOOSE_CATEGORY_MESSAGE_KEY = 'translation:chooseCategoryMessage';

export const TRANSLATION_CHOOSE_SERVICE_MESSAGE_KEY = 'translation:chooseServiceMessage';

export const TRANSLATION_DETAILS_KEY = 'translation:details';

export const TRANSLATION_ID_KEY = 'translation:id';

export const TRANSLATION_SPANISH_KEY = 'translation:spanish';

export const TRANSLATION_ENGLISH_KEY = 'translation:english';

export const TRANSLATION_LANGUAGE_MESSAGE_KEY = 'translation:languageMessage';

export const TRANSLATION_CUSTOMER_DATA_FORM_MESSAGE_KEY = 'translation:customerDataFormMessage';

export const TRANSLATION_NAME_KEY = 'translation:name';

export const TRANSLATION_CONTINUE_KEY = 'translation:continue';

export const TRANSLATION_WELCOME_TURN_KEY = 'translation:welcomeTurn';

export const TRANSLATION_USERS_IN_QUEUE_KEY = 'translation:usersInQueue';

export const TRANSLATION_PRINT_TICKET_KEY = 'translation:printTicket';

export const TRANSLATION_OPTIONS_KEY = 'translation:options';

export const TRANSLATION_CONFIG_MESSAGE_KEY = 'translation:configMessage';

export const TRANSLATION_RECONNECT_MESSAGE_KEY = 'translation:reconnect';

export const TRANSLATION_RETURN_MESSAGE_KEY = 'translation:return';

export const TRANSLATION_CONNECT_KEY = 'translation:connect';

export const TRANSLATION_APPLICATION_KEY = 'translation:application';

export const TRANSLATION_CONNECTED_MESSAGE_KEY = 'translation:alreadyConnectedMessage';

export const TRANSLATION_SUCCESS_CONNECTION_MESSAGE_KEY = 'translation:successConnectionMessage';

export const TRANSLATION_TO_CONNECT_MESSAGE_KEY = 'translation:toConnectMessage';

export const TRANSLATION_INITIAL_CONNECT_MESSAGE_KEY = 'translation:initialConnetMessage';