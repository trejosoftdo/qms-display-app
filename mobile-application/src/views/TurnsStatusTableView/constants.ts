export const BEING_ATTENDED_STATUS_CODE = "BEING_ATTENDED";

export const TO_BE_ATTENDED_STATUS_CODE = "TO_BE_ATTENDED";