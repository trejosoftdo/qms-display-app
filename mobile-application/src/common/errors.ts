import { CONNECTION_EXPIRED_ERROR_MESSAGE, DEVICE_NOT_CONNECTED_ERROR_MESSAGE } from "./constants";

export const DEVICE_NOT_CONNECTED_ERROR = new Error(DEVICE_NOT_CONNECTED_ERROR_MESSAGE);

export const CONNECTION_EXPIRED_ERROR = new Error(CONNECTION_EXPIRED_ERROR_MESSAGE);
