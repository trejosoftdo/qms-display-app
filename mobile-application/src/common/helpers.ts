import { Platform } from 'react-native';
import { router } from 'expo-router';
import { APP_PROTO, EMPTY_VALUE, ONE_SECOND_MILISECONDS, WEB_PLATFORM } from './constants';

/** 
 * Checks if the platform is web
 * @returns boolean
 */
export const isWeb = (): boolean => Platform.OS === WEB_PLATFORM;

/** 
 * Gets the current time
 * @returns number
 */
export const getCurrentTime = (): number => new Date().getTime();

/**
 * Calculates the expire time
 * 
 * @param  {number} expiresIn
 * @returns number
 */
export const calculateExpireTime = (expiresIn: number): number => getCurrentTime() + expiresIn * ONE_SECOND_MILISECONDS;


/** 
 * Extracs icon name
 * @returns string
 */
export const extractIconName = (iconUrl: string): string => iconUrl.replace(APP_PROTO, EMPTY_VALUE)

/**
 * Navigates to a given path
 * 
 * @param  {string} pathname
 * @param  {any} params
 * @returns void
 */
export const goToPath = (pathname: string, params: Record<string, any> = undefined): void => {
  router.push({
    pathname,
    params,
  });
};

/**
 * Waits for certain time
 * @param  {number} timeout 
 * @returns Promise<any>
 */
export const delay = (timeout: number) => new Promise((resolve) => {
  setTimeout(resolve, timeout);
});

/**
 * Gets the response blob url
 * @param  {Response} response 
 * @returns Promise<string>
 */
export const getResponseBlobUrl = async (response: Response): Promise<string> => {
  const stream = await response.blob();
  return URL.createObjectURL(stream);
};
