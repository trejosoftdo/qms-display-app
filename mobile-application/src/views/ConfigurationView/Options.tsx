
import * as React from 'react';
import { StyleSheet } from 'react-native';
import { List } from 'react-native-paper';
import { useTranslation } from 'react-i18next';
import { CONFIG_OPTIONS } from '../../common/constants';
import { TRANSLATION_CONNECTION_KEY, TRANSLATION_INFORMATION_KEY, TRANSLATION_LANGUAGE_KEY, TRANSLATION_OPTIONS_KEY } from '../../common/translations/translation-keys';
import Option from './Option';

/**
 * OptionsProps defines the props for the Options Component.
 *
 * @property onOptionSelect - option selection handler
 */
interface OptionsProps {
  onOptionSelect: (name: string) => void;
};

/**
 * A component a configuration options
 *
 * @param {OptionsProps} props - The props for the Options component.
 */
const Options: React.FC<OptionsProps> = (props: OptionsProps) => {
  const { t } = useTranslation();
  return (
    <List.Section style={styles.container}>
      <List.Subheader>{t(TRANSLATION_OPTIONS_KEY)}</List.Subheader>
      <Option
        title={t(TRANSLATION_INFORMATION_KEY)}
        icon="information"
        onPress={() => props.onOptionSelect(CONFIG_OPTIONS.INFORMATION)}
      />
      <Option
        title={t(TRANSLATION_CONNECTION_KEY)}
        icon="connection"
        onPress={() => props.onOptionSelect(CONFIG_OPTIONS.CONNECTION)}
      />
      <Option
        title={t(TRANSLATION_LANGUAGE_KEY)}
        icon="translate"
        onPress={() => props.onOptionSelect(CONFIG_OPTIONS.LANGUAGE)}
      />
    </List.Section>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    maxWidth: 500,
  },
});

export default Options;