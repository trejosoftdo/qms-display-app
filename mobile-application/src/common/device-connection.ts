import { v4 as uuidv4 } from 'uuid';
import { AuthHeaders, ConnectionDetails } from './models';
import { CONNECTION_DETAILS_KEY, DEVICE_ID_KEY } from './constants';
import { DEVICE_NOT_CONNECTED_ERROR, CONNECTION_EXPIRED_ERROR } from './errors';
import { getValue, setValue } from './store';
import { getNewAccessToken } from './services/auth';
import { BEARER_PORTION } from './constants';
import { getCurrentTime } from './helpers';


/**
 * Checks if a time has expired
 * 
 * @param  {{expiresAt:number}} {expiresAt}
 * @returns boolean
 */
const hasExpired = ({ expiresAt }: { expiresAt: number }): boolean => getCurrentTime() > expiresAt;

/**
 * Gets the device/installation id
 * @returns Promise<string>
 */
export const getDeviceId = async (): Promise<string> => {
  const fetchUUID = await getValue(DEVICE_ID_KEY);
  const deviceId = fetchUUID || uuidv4();
  await setValue(DEVICE_ID_KEY, deviceId.toString());
  return deviceId;
};

/**
 * Saves the connection details
 * 
 * @param  {ConnectionDetails} connectionDetails
 * @returns Promise<void>
 */
export const saveConnectionDetails = async (connectionDetails: ConnectionDetails): Promise<void> => {
  await setValue(CONNECTION_DETAILS_KEY, JSON.stringify(connectionDetails));
};

/**
 * Gets the stored connection details
 * 
 * @returns Promise<ConnectionDetails | undefined>
 */
export const getConnectionDetails = async (): Promise<ConnectionDetails | undefined> => {
  try {
    const data = await getValue(CONNECTION_DETAILS_KEY);
    const details = JSON.parse(data);
    return details as ConnectionDetails;
  } catch (error) {
    return undefined;
  }
};

/**
 * Gets the device auth headers to consume the APIs
 * 
 * @returns AuthHeaders
 */
export const getDeviceAuthHeaders = async (): Promise<AuthHeaders> => {
  const connectionDetails = await getConnectionDetails();

  if (!(connectionDetails?.applicationId && connectionDetails?.accessToken)) {
    throw DEVICE_NOT_CONNECTED_ERROR;
  }

  if (hasExpired(connectionDetails.accessToken)) {
    if (hasExpired(connectionDetails.refreshToken)) {
      throw CONNECTION_EXPIRED_ERROR;
    }

    connectionDetails.accessToken = await getNewAccessToken(
      connectionDetails.applicationId,
      connectionDetails.refreshToken.value,
    );
    await saveConnectionDetails(connectionDetails);
  }

  return {
    applicationId: connectionDetails.applicationId,
    authorization: `${BEARER_PORTION} ${connectionDetails.accessToken.value}`,
  };
};
