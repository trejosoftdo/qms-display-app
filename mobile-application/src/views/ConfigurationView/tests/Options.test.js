import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { TRANSLATION_CONNECTION_KEY, TRANSLATION_INFORMATION_KEY, TRANSLATION_LANGUAGE_KEY } from '../../../common/translations/translation-keys';
import { CONFIG_OPTIONS } from '../../../common/constants';
import Options from '../Options';



describe('Options component', () => {
  const mockProps = {
    onOptionSelect: jest.fn(),
  };

  afterEach(() => {
    jest.clearAllMocks();
  })

  it('renders the component as expected', () => {
    expect(renderShallow(<Options {...mockProps} />)).toMatchSnapshot();
    expect(mockProps.onOptionSelect).toHaveBeenCalledTimes(0);
  });

  it('calls the onOptionSelect handler with information option when selecting it', () => {
    const { getByText } = render(<Options {...mockProps} />);
    const element = getByText(`Translated[${TRANSLATION_INFORMATION_KEY}]({})`);

    fireEvent.press(element);

    expect(mockProps.onOptionSelect).toHaveBeenCalledTimes(1);
    expect(mockProps.onOptionSelect).toHaveBeenCalledWith(CONFIG_OPTIONS.INFORMATION);
  });

  it('calls the onOptionSelect handler with connection option when selecting it', () => {
    const { getByText } = render(<Options {...mockProps} />);
    const element = getByText(`Translated[${TRANSLATION_CONNECTION_KEY}]({})`);

    fireEvent.press(element);

    expect(mockProps.onOptionSelect).toHaveBeenCalledTimes(1);
    expect(mockProps.onOptionSelect).toHaveBeenCalledWith(CONFIG_OPTIONS.CONNECTION);
  });

  it('calls the onOptionSelect handler with language option when selecting it', () => {
    const { getByText } = render(<Options {...mockProps} />);
    const element = getByText(`Translated[${TRANSLATION_LANGUAGE_KEY}]({})`);

    fireEvent.press(element);

    expect(mockProps.onOptionSelect).toHaveBeenCalledTimes(1);
    expect(mockProps.onOptionSelect).toHaveBeenCalledWith(CONFIG_OPTIONS.LANGUAGE);
  });
});
