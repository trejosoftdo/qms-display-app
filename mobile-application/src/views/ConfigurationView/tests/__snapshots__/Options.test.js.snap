// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options component renders the component as expected 1`] = `
<List.Section
  style={
    {
      "maxWidth": 500,
      "width": "100%",
    }
  }
>
  <List.Subheader>
    Translated[translation:options]({})
  </List.Subheader>
  <Option
    icon="information"
    onPress={[Function]}
    title="Translated[translation:information]({})"
  />
  <Option
    icon="connection"
    onPress={[Function]}
    title="Translated[translation:connection]({})"
  />
  <Option
    icon="translate"
    onPress={[Function]}
    title="Translated[translation:language]({})"
  />
</List.Section>
`;
